### PR TITLE
GPU-compatible platypus

### PR DESCRIPTION
**Pull request by Heinrich-BR**
This PR includes a script that builds all of the Platypus dependencies for the Ampere nodes on CSD3 using spack. It also adds to the MFEMProblem class a member which allows the user to choose which device to use in running a simulation. As long as the dependencies have been built with accelerator awareness (which the script does), it should run with GPU acceleration.

I have also added a variable in the `diffusion_mfem.i` test which allows the user to define the device at runtime. So far, the tested options for the `device` variable are `device = cpu` and `device = cuda`.

### CODE REVIEW

File: scripts/build-platypus-csd3-ampere.sh [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+        # shellcheck source=/dev/null
+        . spack/share/spack/setup-env.sh
+    fi
+
+}
+
+install_spack_deps() {
+
+    # Cleaning up everything to start with a new environment
+    spack uninstall -ay arch=${ARCH}
```
Comment (alexanderianblair) [inline scripts/build-platypus-csd3-ampere.sh:75]: This doesn't appear to clean up everything (oddly) - I needed a `spack clean -ab` here as well to resolve this.

Comment (Heinrich-BR) [reply scripts/build-platypus-csd3-ampere.sh:75]: Alright, I've added it!

File: scripts/build-platypus-csd3-ampere.sh [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    spack clean -ab
+
+    spack external find cuda@11.7.1
```
Comment (alexanderianblair) [inline scripts/build-platypus-csd3-ampere.sh:78]: This failed for me - although `spack external find cuda` worked; looks like the version it picked up was 11.7.99 instead. Built OK with 11.7.99 though.

File: src/problem/MFEMProblem.C [hunk 3]
```diff
@@ -23,6 +24,9 @@ MFEMProblem::MFEMProblem(const InputParameters & params)
     _outputs(),
     _exec_params()
 {
+  // logger.set_level(spdlog::level::info);
```
Comment (alexanderianblair) [inline src/problem/MFEMProblem.C:27]: Logger line here should be removed. In MOOSE we can take advantage of the existing informational messaging system [(docs)](https://mooseframework.inl.gov/framework_development/error_warning_messaging.html) and use `mooseInfo(<args>)` for info logs.

Comment (Heinrich-BR) [reply src/problem/MFEMProblem.C:27]: Will do. There are also other commented-out spdlog lines scattered around the code. I'll take the opportunity to remove them too, if that's alright

File: scripts/build-platypus-csd3-ampere.sh [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+#SBATCH --output=platypus_gpu_build.%j.out
+#SBATCH --error=platypus_gpu_build.%j.err
+
+## WARNING: THIS SCRIPT WILL UNINSTALL ALL SPACK MODULES ASSOCIATED WITH
+## THE ARCHITECTURE DEFINED IN THE ARCH VARIABLE. IF YOU DO NOT WISH TO DO
+## THAT, COMMENT OUT THE SPACK UNINSTALL LINE BEFORE SUBMITTING THE SCRIPT
+## -> UNINSTALL LINE IN THE install_spack_deps() FUNCTION
+ARCH="linux-rocky8-zen"
+
+export compile_cores=32
```
[the remark this answers is not in the document]
Comment (alexanderianblair) [reply scripts/build-platypus-csd3-ampere.sh:19]: Setting `--cpus-per-task=32` now appears to be working - must have been an issue on CSD3's end. Either way, we should avoid oversubscription

File: scripts/build-platypus-csd3-ampere.sh [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+        -DBUILD_SHARED_LIBS=YES \
+        -DMFEM_USE_OPENMP=NO \
+        -DMFEM_THREAD_SAFE=YES \
+        -DMFEM_ENABLE_EXAMPLES=YES \
+        -DMFEM_ENABLE_MINIAPPS=YES \
+        -DMFEM_USE_MPI=YES \
+        -DMFEM_USE_CUDA=YES \
+        -DCUDA_ARCH=sm_80 \
+        -DMFEM_USE_METIS_5=YES \
+        -DMFEM_USE_SUPERLU=YES \
```
Comment (alexanderianblair) [inline scripts/build-platypus-csd3-ampere.sh:167]: The new `build-platypus-csd3-ampere.sh` build script is currently failing for me on a new build when building MFEM:

> -- SuperLUDist: looking for required package: BLAS
-- Performing Test SuperLUDist_VERSION_OK
-- Performing Test SuperLUDist_VERSION_OK - Failed
-- Configuring incomplete, errors occurred!
MFEM build failed

File: scripts/build-platypus-csd3-ampere.sh [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+    echo "Building platypus..."
+    git clone https://github.com/aurora-multiphysics/platypus.git
+    cd platypus || exit 1
+    git submodule update --init --recursive
+    cd contrib/hephaestus/ || exit 1
+    mkdir build
+    cd build || exit 1
+    cmake -G Ninja -DCMAKE_BUILD_TYPE=Release -DMFEM_DIR="${BUILD_PATH}/mfem/build" ..
+    ninja
+    cd "${BUILD_PATH}"/platypus || exit 1
```
Comment (alexanderianblair) [inline scripts/build-platypus-csd3-ampere.sh:262]: The `hephaestus` submodule no longer exists in `platypus` - this section causes an error, as it attempts to `cd` into a non-existent directory. Should just build platypus here.

File: scripts/build-platypus-csd3-ampere.sh (hunk 5 above)
Comment (alexanderianblair) [reply scripts/build-platypus-csd3-ampere.sh:169]: Resolved by addition of new SuperLU_Dist CMAKE flags

**Review by alexanderianblair — APPROVED** [review]
I've checked the build script runs and builds a GPU-compatible Platypus version successfully on CSD3 - looks OK this end.

**Pull request merged**